### DISCUSSION TITLE
Does `viem` supports `custom solidity errors`?

### DISCUSSION THREAD

**Discussion by kostysh** [wevm/viem#589] [Question] [answered]
When calling a contract I am getting an error without any revert reasons:

Here is an example:

```
ContractFunctionExecutionError: An internal error was received.

Raw Call Arguments:
  from:  // address
  to:    // address
  data:  // call data
 
Contract Call:
  address:   // address
  function:  // function name
  args:       // args
  sender:   // address

Docs: https://viem.sh/docs/contract/simulateContract.html
Details: Internal JSON-RPC error.
Version: viem@0.3.36
```

Meanwhile, In the Javascript console I see that the Metamask returns the following JSON RPC error:

```json
{
    "code": -32603,
    "message": "Internal JSON-RPC error.",
    "data": {
        "code": 3,
        "message": "execution reverted",
        "data": "0xa2fb840f"
    }
}
```

Yes, I can identify an error by iterating through the errors ABI and comparing encoded error selectors (_data.message === keccak256("MyCustomError()").substr(0, 10)_). 

But, `simulateContract` does not return a JSON RPC error in the return. I can see this error only in the console.

Is it possible to handle custom solidity errors? If yes, then how. Thanks.

**Comment by jxom** [comment]
Yes, it is supported and it should be automatic. We have a test for it here: https://github.com/wagmi-dev/viem/blob/main/src/actions/public/simulateContract.test.ts#L242

Might be something to do with the response structure. Will take a look.

Off-topic: I would avoid using `window.ethereum` for simulating contracts as not all wallets support `eth_call`. But if you know your consumers are only using MetaMask, then it should be fine.

**Reply by kostysh** [reply]
> Off-topic: I would avoid using `window.ethereum` for simulating contracts as not all wallets support `eth_call`. But if you know your consumers are only using MetaMask, then it should be fine.

sure, agree. I will be using `http` transport for the public client

**Reply by kostysh** [reply, marked as answer]
I can confirm that using `http` transport for the public client has fixed the issue. Now `cause.data.errorName` returns a custom error
